kernel/copy.masm

```diff
@@ -0,0 +1,256 @@
+  sub
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, writePtr, words_even, needs_padding, ...]
+  padw padw
+
+  # stack: [A, B, C, writePtr, needs_padding, ...]
+  exec.move_from_tape_to_memory_even_words
+
+  # stack: [needs_padding, A, B, C, writePtr, ...]
+  movup.13
+
+  # if(needs_padding)
+  if.true 
+    # Rescue Prime Optimized uses overwrite mode. Drop `A`.
+    # stack: [B, C, writePtr, ...]
+    dropw
+
+    # Overwrite the `B` with the new data
+    # stack: [B', C, writePtr, ...]
+    adv_loadw
+
+    # - get the memory address that B' should be saved to
+    # - update the writePtr to point to the next address
+    # stack: [writePtr, B', writePtr+1, C, ...]
+    movup.8 dup.0 add.1 movdn.5
+
+    # save data to memory
+    # stack: [B', writePtr+1, C, ...]
+    mem_storew
+
+    # Fix writePtr position
+    # stack: [B', C, writePtr+1, ...]
+    movup.4 movdn.8
+
+    # Push padding word
+    # stack: [A, B', C, writePtr+1, ...]
+    push.1.0.0.0
+
+    # Run RPO permutation
+    # stack: [A', B', C', writePtr+1, ...]
+    rpperm
+  end
+
+  # The RPO hash is a single word, discard the unused one and the capacity.
+  # stack: [RpoHash, writePtr', ...]
+  dropw swapw dropw
+end
+
+# move_from_tape_to_memory_with_commitment(words: Felt, writePtr: Address, commitment: Commitment) -> [writePtr']
+#
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, at the end asserts the commitment data was copied
+proc.move_from_tape_to_memory_with_commitment.0
+  # Copies the tape data to memory
+  # stack: [RpoHash, writePtr', Commitment, ...]
+  exec.move_from_tape_to_memory
+
+  # Save the writePtr
+  # stack: [RpoHash, Commitment, writePtr', ...]
+  movup.4 movdn.8
+
+  # Check the commitment
+  # stack: [writePtr', ...]
+  exec.assert_eqw
+end
+
+# memcopy(readPtr: Address, writePtr: Address, words: Felt)
+#
+# Copies `words` words from `readPtr` to `writePtr`.
+proc.memcopy.0
+  # input: [readPtr, writePtr, words, ...]
+  # output: [...]
+
+  # LOOP: [readPtr, writePtr, words, ...]
+  # while(words>0)
+  dup.3 neq.0 while.true
+    # prepare read
+    # stack: [readPtr, 0, 0, 0, 0, readPtr, writePtr, words, ...]
+    padw dup.4
+
+    # perform read
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_loadw
+
+    # prepare write
+    # stack: [writePtr, A, readPtr, writePtr, words, ...]
+    dup.5
+
+    # peform write
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_storew
+
+    # clean stack
+    # stack: [readPtr, writePtr, words, ...]
+    dropw
+
+    # update counters
+    # LOOP: [readPtr+1, writePtr+1, words-1, ...]
+    inc.1 movup.2 sub.1 movup.2 inc.1 movup.2
+
+    dup.3 neq.0 # while(words>0)
+  end
```

Comment:
A couple of notes here:

1. I'd avoid pushing and popping values here as much as possible - this will contribute a lot to the cycle count. Instead, we can use 4 elements of the stack as a buffer into which we first `mem_loadw` and then `mem_storew`. We don't need to use `padw` and `dropw` in between these.
2. For the counters, it might be more efficient to keep the target number and count up until the count reaches the target number (rather than counting to 0). This is because incrementing by 1 is cheap (1 VM cycles) but subtracting 1 is not as cheap (3 VM cycles). Also, I don't think `inc.1` is a valid instruction - you can use `add.1` instead.

Reply:
just for reference, this discussion continued on the miden-vm PR https://github.com/0xPolygonMiden/miden-vm/pull/668